src/Npgsql/Internal/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -103,30 +124,116 @@ public override decimal Read(NpgsqlReadBuffer buf, int len, FieldDescription? fi
+            }
+
+            var digits = new ushort[groups];
+
+            for (var i = 0; i < groups; i++)
+            {
+                await buf.Ensure(sizeof(ushort), async);
+                digits[i] = buf.ReadUInt16();
+            }
+
+            // Calculate powers 10^8, 10^16, 10^32, ...
+            // We should have the last calculated power to be less than the input
+            var lenPow = 2; // 2 ushorts fit in one uint, represents 10^8
+            var numPowers = 0;
+            while (lenPow < weightLeft + 1)
+            {
+                lenPow <<= 1;
+                ++numPowers;
+            }
+            var factors = numPowers > 0 ? new BigInteger[numPowers] : null;
+            if (numPowers > 0)
+            {
+                factors![0] = new BigInteger(100000000U);
+                for (var i = 1; i < numPowers; i++)
+                    factors[i] = factors[i - 1] * factors[i - 1];
+            }
+
+            var result = ToBigIntegerInner(0, weightLeft + 1, digits, factors);
+            return sign == SignPositive ? result : -result;
+
+            static BigInteger ToBigIntegerInner(int offset, int length, ushort[] digits, BigInteger[]? factors)
+            {
+                if (length <= 2)
+                {
+                    var r = 0U;
+                    for (var i = offset; i < offset + length; i++)
+                    {
+                        r *= 10000U;
+                        r += i < digits.Length ? digits[i] : 0U;
+                    }
+                    return r;
+                }
+                else
+                {
+                    // Split the input into two halves, the lower one should be a power of two in digit length,
+                    // then multiply the higher part with a precomputed power of 10^8 and add the results.
+                    var lenFirstHalf = 2 << 1; // 2 ushorts fit in one uint, skip 1 since we've already covered the base case.
+                    var pos = 0;
+                    while (lenFirstHalf < length)
+                    {
+                        lenFirstHalf <<= 1;
+                        ++pos;
+                    }
+                    var factor = factors![pos];
+                    lenFirstHalf >>= 1;
+                    var lo = ToBigIntegerInner(offset + length - lenFirstHalf, lenFirstHalf, digits, factors);
+                    var hi = ToBigIntegerInner(offset, length - lenFirstHalf, digits, factors);
+                    return hi * factor + lo; // .NET uses Karatsuba multiplication, so this will be fast.
+                }
+            }
+        }
 
         #endregion
 
         #region Write
 
         /// <inheritdoc />
-        public override int ValidateAndGetLength(decimal value, NpgsqlParameter? parameter)
+        public override int ValidateAndGetLength(decimal value, ref NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter)
```

Comment:
Currently the code is a bit redundant, the validation part and the actual writing part do more or less the same math. Should we use the `parameter.ConvertedValue` here, to trade duplicade code and cpu time against an extra memory allocation?

Reply:
We definitely can, though note that ConvertedValue only works for top-level parameters; if we're writing an array of numerics, that won't be used (at some point we can ideally extend the infrastructure, though I was although thinking about switching to one-pass writing instead...).

(for duplicate code, we can maybe refactor out a common method or something...)

Reply:
How would one-pass writing work? We need to know the length before serializing the value.
Anyway, maybe we can generalize the "length cache" to also be able to store converted values? Then we can get rid of the parameter's `ConvertedValue`. This will also then work for arrays and composite types containing for example a BigInteger.

Reply:
> How would one-pass writing work? We need to know the length before serializing the value.

I posted some thoughts and played around a couple years ago: #1727. Didn't see anything at the time that made an interesting perf difference, but this should probably be revisited at some point.

> Anyway, maybe we can generalize the "length cache" to also be able to store converted values?

Yep, that's definitely a possibility - we could make it a more generic cache. Though I'm not sure there are many mainstream scenarios where this would improve things significantly - an array of huge BigDecimals is probably going to be... rare...